### PR TITLE
feat(swc): Add a new source maps option "linked"

### PR DESCRIPTION
**Pull request by realtimetodie**
Adds a new source maps option "linked" to generate a source map file and append a sourceMappingURL comment to the end of the output, containing an URL to the generated source map file.

This is only implemented in `@swc/cli`

https://github.com/swc-project/cli/blob/933589ae907637afb7f0b0a43b7a24a7a920595e/src/swc/compile.ts#L31

Possible values for the --source-maps option

- **false** (default): Disables the source map generation.
- **true**: Generates a source map file, but without to append a sourceMappingURL comment to the end of the output.
- **inline**: Generates a source map file and appends a sourceMappingURL comment to the end of the output, containing the base64 encoded source map.

    ```
    //# sourceMappingURL=data:application/json;base64,<encoded_source_map_data>
    ```

New

- **linked**: Generates a source map file and appends a sourceMappingURL comment to the end of the output, containing an URL to the generated source map file.

    ```
    //# sourceMappingURL=<source_map_url>
    ```

Example

```bash
$ swc compile input.js --out-file=output.js --source-maps=linked
```

Output

```
output.js
output.js.map
```

Show the last line of the generated output.js file

```
$ tail -n=1 output.js
//# sourceMappingURL=output.js.map
```

Summary

- The source map path is now known to the compiler.
- Replaced the enum ~~`SourceMapsConfig::Str`~~ with the enum `SourceMapsConfig::Inline` and the enum `SourceMapsConfig::Linked`.
- Replaced the enum ~~`InputSourceMap::Str`~~ with the enum `InputSourceMap::Inline`.
- `SourceMapsConfig::default()` is now set to `SourceMapsConfig::Bool(false)` to match the compiler default (source map generation is disabled by default).
- Added the `FromStr` trait to `SourceMapsConfig`.
- Added the `serde::Deserializer` trait to `SourceMapsConfig`.
- Added the `serde::Deserializer` trait to `InputSourceMap`.
- Added tests
- Improved the error message when the `SourceMapsConfig::from_str` method is used

    Example

    ```bash
    $ swc compile input.js --out-file=output.js --source-maps=foo
    ```

    Error message

    ```
    invalid source maps value: got "foo" but expected either one of false, true, inline, linked
    ```

Related

- https://github.com/swc-project/swc/discussions/3859
- https://github.com/swc-project/swc/discussions/6479
- https://github.com/aspect-build/rules_swc/issues/146

### CODE REVIEW

**Comment by jridgewell** [comment]
This breaks with both [Babel](https://babeljs.io/docs/en/options#sourcemaps)'s and [esbuild](https://esbuild.github.io/api/#sourcemap)'s definition of `"both"`, which is to add an a `//# sourceMappingURL=data:application/json;base64,…` and have an external sourcemap file. Babel doesn't have an option like what's proposed here (probably because it doesn't' bundle), but esbuild calls this `"linked"`

**Comment by realtimetodie** [comment]
Thank you @jridgewell. This is a good idea, I like it.

I renamed the enum ~~`SourceMapsConfig::Both`~~ ("both") -> `SourceMapsConfig::Linked` ("linked") to not break with [Babel](https://babeljs.io/docs/en/options#sourcemaps)'s and [esbuild](https://esbuild.github.io/api/#sourcemap)'s definition of "both" as suggested.

Example

```bash
$ swc compile input.js --out-file=output.js --source-maps=linked
```

**Comment by realtimetodie** [comment]
I updated the title and the initial comment.

**Comment by kdy1** [comment]
> **true**: Generates a source map file, but without to append a sourceMappingURL comment to the end of the output

Is this true? I think this is a bug..

**Review by kdy1 — CHANGES_REQUESTED** [review]
Please use `true` for this instead

**Comment by kdy1** [comment]
Nevermind. I concluded `linked` is fine.

File: bindings/binding_core_node/Cargo.toml
Comment (kdy1) [inline bindings/binding_core_node/Cargo.toml:59]: Please revert this

Comment (kdy1) [reply bindings/binding_core_node/Cargo.toml:59]: I think we should have only `verison = "0.56.2"` (latest)

**Comment by realtimetodie** [comment]
@kdy1 I'm not sure why the bindings are in an isolated workspace. It complicates the development.

**Review by jridgewell — DISMISSED** [review]

**Comment by alexeagle** [comment]
@realtimetodie do you think that build failure for the node binding is caused by this change?
```
error[E0061]: this function takes 12 arguments but 13 arguments were supplied
   --> binding_core_node/src/bundle.rs:127:51
    |
127 | ...                   let output = self.swc.print(
    |                                             ^^^^^
    |
note: associated function defined here
   --> /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/swc-0.242.2/src/lib.rs:438:12
    |
438 |     pub fn print<T>(
    |            ^^^^^
help: did you mean
    |
127 |                             let output = self.swc.print(&m, None, None, true, codegen_target, SourceMapsConfig::Bool(true), &Default::default(), None, minify, None, false, true)?;
```

**Comment by realtimetodie** [comment]
@alexeagle Yes, e4d80362e12c8c718c65957efad4a448c16e44d8 broke the build. Cargo pulls the swc_core crate from the remote crate registry now, which does not contain the new changes. Before that, the build passed. This is what I meant in https://github.com/swc-project/swc/pull/6817#issuecomment-1383645651. I think @kdy1 will bump the version manually, once this is merged.

**Comment by alexeagle** [comment]
Ping @kdy1 - we're waiting for this one so we can release the Bazel swc integration to 1.0.0

**Comment by kdy1** [comment]
There's a merge conflict

**Comment by realtimetodie** [comment]
I will resolve the merge conflict in a couple of hours, I've a night shift today in the warehouse

**Comment by realtimetodie** [comment]
I resolved the merge conflict. The version of the swc_core crate is now set to 0.58.0.

File: crates/swc/src/config/mod.rs
Comment (kdy1) [inline crates/swc/src/config/mod.rs:252]: I think this should be `Linked(PathBuf)` instead

Comment (kdy1) [reply crates/swc/src/config/mod.rs:252]: `Linked` + `None` for `source_map_path` is an invalid state

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc/src/config/mod.rs
Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:252]: Yes, I also thought about this. I want to emphasize that this will not solve the dependent-type problem.

For example, the instantiation creates an invalid Path to a nonexistent source maps file already

```json
{
    "sourceMaps": "linked"
}
```

```rust
SourceMapsConfig::Linked(PathBuf::new())
```

The value of the `SourceMapsConfig::Linked` enum can not be universal. Its inner `PathBuf` value (which is the path to the source maps file) is dynamic and changes with every input.

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:252]: I tried this out and the code is even more ugly now.

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:252]: To get this right, the `SourceMapsConfig::Bool(bool)` needs to be replaced with `SourceMapsConfig::Enabled(PathBuf)` and `SourceMapsConfig::Disabled` respectively. Please let me know if I this is alright.

Comment (kdy1) [reply crates/swc/src/config/mod.rs:252]: Why do we need pathbuf as an argument? `true` simply generates `.map` file with same file name, so I think we should do same thing for `linked`, too

**Comment by realtimetodie** [comment]
I had to delete the fixture test for the linked option due to the way the test runner works. It is not possible to construct a source map path from a .swcrc configuration alone, as the initial value is `PathBuf::new()`.

File: crates/swc/src/config/mod.rs
Comment (realtimetodie) [reply crates/swc/src/config/mod.rs]: The reason is that I would prefer to move entire generation of the .map path into the swc core crate. If we would do that, then we could also return a map path from there.

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs]: I implemented it as you suggested, but I had to delete the fixture test for the linked option.

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs]: I implemented it as you suggested, but I had to delete the fixture test for the linked option. This is the down-side of your idea.

File: crates/swc/src/config/mod.rs
Comment (kdy1) [inline crates/swc/src/config/mod.rs:249]: I think this should be `Linked`, not `Linked(PathBuf)`

**Review by kdy1 — CHANGES_REQUESTED** [review]
I was confused while reviewing for the first time, my bad.

Please see https://github.com/swc-project/swc/pull/6817#discussion_r1083376046

File: crates/swc/src/config/mod.rs
Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:249]: No problem. I will revert my previous commit.

File: crates/swc/tests/exec.rs
Comment (kdy1) [inline crates/swc/tests/exec.rs:160]: Please revert this. The execution test suite is very slow and adding a matrix will prolong it

File: bindings/swc_cli/src/commands/compile.rs
Comment (kdy1) [inline bindings/swc_cli/src/commands/compile.rs:461]: Seems like a mistake?

Comment (realtimetodie) [reply bindings/swc_cli/src/commands/compile.rs:461]: Previously, the name of the source map was defined in the aforementioned function. But this does not work anymore. It is now required to know the name of the source map beforehand. Therefore, I moved this function and made it part of the executor.

**Comment by realtimetodie** [comment]
I dropped 233b6746303b24b4add6aeb81a14340c2f02149e and reverted the changes.

File: bindings/swc_cli/src/commands/compile.rs
Comment (realtimetodie) [reply bindings/swc_cli/src/commands/compile.rs:459]: The function still exists, it is just inside of the executor.

**Comment by realtimetodie** [comment]
@kdy I want to ask you politely if anything is missing, please let me know. Thank you.

File: crates/swc/src/config/mod.rs
Comment (kdy1) [inline crates/swc/src/config/mod.rs:1278]: I'm not merging this PR because of this

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:1278]: I know that this is a large PR and I understand that you are doing this voluntarily. I apologize for the inconvenience.

Comment (kdy1) [reply crates/swc/src/config/mod.rs:1278]: No, I'm talking about adding this field. I meant removing this field by `implicit` path.

https://github.com/swc-project/swc/pull/6817#discussion_r1083376046

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs:1278]: I can't follow. It would be nice if you could provide me a more detailed example. Thank you.

Comment (kdy1) [inline crates/swc/src/config/mod.rs:1278]:
```suggestion

```

File: crates/swc/src/config/mod.rs
Comment (kdy1) [inline crates/swc/src/config/mod.rs:1277]:
```suggestion

```

Comment (realtimetodie) [reply crates/swc/src/config/mod.rs]: I did exactly as you said. Thank you for your explanation.

**Comment by realtimetodie** [comment]
I invested a lot of hours to help https://github.com/swc-project/swc/discussions/3859. I do this voluntarily and no one is paying me for doing this. The build is now broken.

**Comment by realtimetodie** [comment]
@paullewis @alexeagle Please help me I think @kdy1 is playing games with me here.

**Comment by realtimetodie** [comment]
I was nice and accommodating. I don't know @kdy1. But the way he reviewed my work and communicated with me is rude and disrespectful.

**Comment by kdy1** [comment]
@realtimetodie  I told we should preserve the previous behavior (implicitly determining the path of the `.map` file) you several times. It's you ignoring my comment. I attached the link to my previous comment each time I request a change.

Of course, it's possible that someone fails to understand my change request. I reviewed an enormous amount of PRs until now and I have lots of experience like it. In my experience, they understood if I showed code to them. So I show you the code change for my request using the GitHub feature. Game? Rude? For which point?

**Comment by realtimetodie** [comment]
I think what you mean with "implicitly determined" is to generate the source maps path within the compiler from.the filename, instead of passing it down as an argument? 

The reason why we can't "implicitly determine" the path to the source map file is because of the cli --source-map-target option. 

Currently, the user can define a custom output path for the generated source maps file when using the SourceMapsConfig::Bool(true) enum. I simply adopted this feature for the new SourceMapsConfig::Linked enum too. This is the only reason why the source_maps_path is passed down to the compiler.

If you say that the cli --source-map-target option should be ignored let me know. I thought you saw this part in the code.

**Comment by kdy1** [comment]
> I think what you mean with "implicitly determined" is to generate the source maps path within the compiler from.the filename, instead of passing it down as an argument?

Yep. I mean filename + `.map`, which is the previous logic.

But seems like it's my bad, sorry. I didn't know that `--source-map-target` exist. I'm not sure how did previous code work

**Comment by kdy1** [comment]
Ah, I got it. It simply returned the `map` object. But with `linked` it's impossible because `linked` requires writing the output path to the source file, right?

I'm sorry, I didn't think about this carefully. (I stopped thinking to work on stc while thinking about the source map path option)

**Comment by realtimetodie** [comment]
It is also my fault, I did not point this out in the initial comment when i listed the changes and features.

**Review by kdy1 — DISMISSED** [review]
Thank you, and sorry.

---

swc-bump:
 - swc --breaking

**Comment by realtimetodie** [comment]
I deleted my previous comments. I was too emotional.

**Comment by kdy1** [comment]
Oh CI failed

**Comment by realtimetodie** [comment]
I need to revert https://github.com/swc-project/swc/pull/6817/commits/4b303a27853c7434716dc2a5c1d804c30bbcf1f1.

**Review by kdy1 — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]

**Comment by kdy1** [comment]
cc @kwonoj I think this can also be implemented by adding logic like https://github.com/swc-project/cli/blob/933589ae907637afb7f0b0a43b7a24a7a920595e/src/swc/compile.ts#L30-L33 in `swc_cli`.
Which way do you prefer?

**Comment by kdy1** [comment]
I thought more about this, and I think sourceMap=true should generate `sourceMappingUrl`, and `linked` seems a bit redundant. But this can't be done by the `swc` crate.

**Comment by realtimetodie** [comment]
I apologize for not writing my thoughts about the implementationin the initial post.

The main reason why I put the generation of the `sourceMappingURL` string into the swc crate is because I favor centralization. The code to generate the `sourceMappingURL` would be in one place, in the `print` function and would not be scattered around.

Secondly, I thought that third party consumers of the swc crate would greatly benefit from such a new feature SourceMapsConfig::Linked. Otherwise the feature would be hidden in the cli bindings crate, which is not meant to be consumed.

I understand the concerns about the new `source_map_path` variable. It seems unfortunate at first, but the path to the generated output is passed down using the `output_path` variable already.

https://github.com/swc-project/swc/blob/33dab568531fa3fc77697b75bce2df7514938d0b/crates/swc/src/lib.rs#L438-L442

I also want to emphasize on @jridgewell https://github.com/swc-project/swc/pull/6817#issuecomment-1383230637. The introduction of a new "linked" option helps users, because the configuration and naming convention will be aligned with other tools.

**Comment by kdy1** [comment]
I thought more about this. I think an ideal solution is use `SourceMapsConfig::Bool(true)` for this behavior and patch  `@swc/cli` not to add `sourceMappingUrl`.

The problem is that, `@swc/cli` 

 - generates `sourceMappingUrl` for `SourceMapsConfig::Bool(true)`
 - generates two `sourceMappingUrl`s for `SourceMapsConfig::Linked`

and `swc_cli` 

 - generates no `sourceMappingUrl` for `SourceMapsConfig::Bool(true)`
 - generates one `sourceMappingUrl` for `SourceMapsConfig::Linked`

**Comment by realtimetodie** [comment]
@kdy1 Yes, the @swc/cli needs to be patched.

However, the Node.js cli can handle multiple inputs at once and concats the outputs. If we combine the SourceMapsConfig::Bool(true) option as suggested, this will create a regression because every generated output will contain an URL to a non-existent source map file

https://github.com/swc-project/cli/blob/7217f7ad6a68f5bc788094affd8c5530231911c5/src/swc/file.ts#L33

For this reason, we will always need an option to generate an output and a source map without a sourceMappingURL string. 

Ideally, we could try to move this part from the Node.js cli into Rust. But this would require a lot of effort.
The new option SourceMapsConfig::Linked requires minimal changes to the Node.js cli and does not break existing behaviour.

**Comment by realtimetodie** [comment]
@kdy1 The @swc/cli can be patched in such a way, that users will not require to update their configuration (from true -> "linked") and the option "linked" would be invalid in Node.js.

**Comment by realtimetodie** [comment]
I linked to https://github.com/swc-project/swc/issues/1388 which is something I will be working on after this PR, as it is a regression of the Rust cli.

**Comment by kdy1** [comment]
I'm against adding an option which is invalid for `@swc/cli`

**Pull request closed without merging**
